format/udf-spec.md

```diff
@@ -0,0 +1,285 @@
+ - Unless required by applicable law or agreed to in writing, software
+ - distributed under the License is distributed on an "AS IS" BASIS,
+ - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ - See the License for the specific language governing permissions and
+ - limitations under the License.
+ -->
+
+# Iceberg SQL UDF Spec
+
+## Background and Motivation
+
+A SQL user-defined function (UDF or UDTF) is a callable routine that accepts input parameters and executes a function body.
+Depending on the function type, the result can be:
+
+- **Scalar functions (UDFs)** – return a scalar value, which may be a primitive type (e.g., `int`, `string`) or a non-primitive type (e.g., `struct`, `list`).
+- **Table functions (UDTFs)** – return a table, i.e., a table with zero or more rows and columns with a uniform schema.
+
+Many compute engines (e.g., Spark, Trino) already support UDFs, but in different and incompatible ways. Without a common
+standard, UDFs cannot be reliably shared across engines or reused in multi-engine environments.
+
+This specification introduces a standardized metadata format for UDFs in Iceberg. 
+
+## Goals
+
+* Define a portable metadata format for both scalar and table SQL UDFs. The metadata is self-contained and can be moved across catalogs. 
+* Support function evolution through versioning and rollback.
+* Provide consistent semantics for representing UDFs across engines.
+
+## Overview
+
+UDF metadata follows the same design principles as Iceberg table and view metadata: each function is represented by a
+**self-contained metadata file**. Metadata captures definitions, parameters, return types, documentation, security,
+properties, and engine-specific representations.
+
+* Any modification (new overload, updated representation, changed properties, etc.) creates a new metadata file, and atomically swaps in the new file as the current metadata.
+* Each metadata file includes recent definition versions, enabling rollbacks without external state.
+
+## Specification
+
+### UDF Metadata
+The UDF metadata file has the following fields:
+
+| Requirement | Field name        | Type                   | Description                                                                                                     |
+|-------------|-------------------|------------------------|-----------------------------------------------------------------------------------------------------------------|
+| *required*  | `function-uuid`   | `string`               | A UUID that identifies the function, generated once at creation.                                                |
+| *required*  | `format-version`  | `int`                  | Metadata format version (must be `1`).                                                                          |
+| *required*  | `definitions`     | `list<overload>`       | List of function [overload](#overload) entities.                                                                |
```

Comment:
These names are a product of the evolution of this spec and I think we are now in a good place to consider choosing more consistent names. It is strange that "definitions" is actually a list of "overloads". I think we should choose a single term for a function signature and use that here. To me, "definition" inside "definitions" makes the most sense.